### PR TITLE
CNS-3192 Common Proxy Settings

### PR DESCRIPTION
**Pull request by rosti**
Allow for common proxy settings to be specified to all components at once by a new section in the CRD.

### CODE REVIEW

File: api/v1/cbcontainersagent_types.go [hunk 1]
```diff
@@ -83,6 +91,39 @@ func (s CBContainersComponentsSettings) ShouldCreateDefaultImagePullSecrets() bo
 	return *s.CreateDefaultImagePullSecrets
 }
 
+type CBContainersProxySettings struct {
```
Comment (Shanik1) [inline api/v1/cbcontainersagent_types.go:94]: Are the attributes must be pointers?

File: cbcontainers/state/common/env_var.go [hunk 2]
```diff
@@ -61,6 +65,44 @@ func (b *EnvVarBuilder) WithCustom(customEnvsToAdd ...coreV1.EnvVar) *EnvVarBuil
 	return b
 }
 
+func (b *EnvVarBuilder) WithProxySettings(proxySettings *cbcontainersv1.CBContainersProxySettings) *EnvVarBuilder {
+	if *proxySettings.Enabled == false {
```
Comment (Shanik1) [inline cbcontainers/state/common/env_var.go:69]: Potential nil pointer reference. If the value is always set (false/true), then this shouldn't be a pointer, and just be a bool

File: cbcontainers/state/components/sensor_daemon_set.go [hunk 3]
```diff
@@ -408,7 +408,8 @@ func (obj *SensorDaemonSetK8sObject) mutateRuntimeEnvVars(container *coreV1.Cont
 
 	envVarBuilder := commonState.NewEnvVarBuilder().
 		WithCustom(customEnvs...).
-		WithSpec(sensorSpec.Env)
+		WithSpec(sensorSpec.Env).
+		WithProxySettings(agentSpec.Components.Settings.Proxy)
```
Comment (Shanik1) [inline cbcontainers/state/components/sensor_daemon_set.go:412]: For the **runtime sensor,** we want to unpack the `cbcontainers-runtime-resolver.cbcontainers-dataplane.svc.cluster.local` NO_PROXY value by default (unless it was already set by the customer). 

That is aim to simplify the customer CRD setting - customers was always forgetting to fill this value (documented [here](https://github.com/octarinesec/octarine-operator#using-http-proxy) ) or filled it incorrectly. That is why we wanted to fill them ourselves in the operator.

File: cbcontainers/state/components/sensor_daemon_set.go [hunk 4]
```diff
@@ -536,7 +538,8 @@ func (obj *SensorDaemonSetK8sObject) mutateClusterScannerEnvVars(container *core
 		WithEnvVarFromResource("CLUSTER_SCANNER_REQUESTS_MEMORY", ClusterScanningContainerName, "requests.memory").
 		WithEnvVarFromField("CLUSTER_SCANNER_NODE_NAME", "spec.nodeName", "v1").
 		WithSpec(clusterScannerSpec.Env).
-		WithGatewayTLS()
+		WithGatewayTLS().
+		WithProxySettings(agentSpec.Components.Settings.Proxy)
```
Comment (Shanik1) [inline cbcontainers/state/components/sensor_daemon_set.go:542]: For the **cluster-scanner**, we want to unpack the `cbcontainers-image-scanning-reporter.cbcontainers-dataplane.svc.cluster.local` NO_PROXY value by default (unless it was already set by the customer). 

That is aim to simplify the customer CRD setting - customers was always forgetting to fill this value (documented [here](https://github.com/octarinesec/octarine-operator#using-http-proxy) ) or filled it incorrectly. That is why we wanted to fill them ourselves in the operator.

File: api/v1/cbcontainersagent_types.go [hunk 5]
```diff
@@ -83,6 +91,39 @@ func (s CBContainersComponentsSettings) ShouldCreateDefaultImagePullSecrets() bo
...
+	// Enabled controls if the proxy settings are applied or not
+	//
+	// +kubebuilder:default:=false
+	Enabled *bool `json:"enabled,omitempty"`
+
+	// HttpProxy points to the URL of the HTTP proxy.
+	// If set, it'll result in an additional HTTP_PROXY environment variable defined for all
+	// components. When a component already has HTTP_PROXY defined through the CRD, HttpProxy won't
+	// be used, using the component's original HTTP_PROXY value instead.
+	HttpProxy *string `json:"httpProxy,omitempty"`
```
Comment (Shanik1) [inline api/v1/cbcontainersagent_types.go:104]: Do we need kubebuilder values for these attributes too? @BenRub

**Comment by ltsonov-cb** [comment]
Don't forget to update the CRD in the `charts` folder (haven't looked at the rest of the MR but I noticed that we changed the CR under `config/bases` and not under the helm charts

Unfortunately that is manual at the moment.

**Review by BenRub — APPROVED** [review]

**Pull request merged**